File: backend/src/middleware/requestId.ts
import { randomUUID } from "crypto";
import type { RequestHandler } from "express";

const MAX_LEN = 128;
const SAFE_INCOMING = /^[a-zA-Z0-9._-]+$/;

function normalizeIncoming(raw: string | undefined): string | null {
  if (raw === undefined) {
    return null;
  }
  const trimmed = raw.trim();
  if (trimmed.length === 0 || trimmed.length > MAX_LEN || !SAFE_INCOMING.test(trimmed)) {
    return null;
  }
  return trimmed;
}

export function assignRequestId(): RequestHandler {
  return (req, res, next) => {
    const id = normalizeIncoming(req.get("x-request-id")) ?? randomUUID();
    req.requestId = id;
    res.setHeader("X-Request-Id", id);
    next();
  };
}

File: python_rag/request_id.py
from __future__ import annotations

import re
import uuid

_MAX_LEN = 128
_SAFE_INCOMING = re.compile(r"^[a-zA-Z0-9._-]+$")


def normalize_incoming(raw: str | None) -> str | None:
    if raw is None:
        return None
    trimmed = raw.strip()
    if not trimmed or len(trimmed) > _MAX_LEN or _SAFE_INCOMING.fullmatch(trimmed) is None:
        return None
    return trimmed


def assign_request_id(header_value: str | None) -> str:
    return normalize_incoming(header_value) or str(uuid.uuid4())
